master/internal/api_experiment.go

```diff
@@ -78,6 +81,72 @@ func (a *apiServer) GetExperiment(
 	return &apiv1.GetExperimentResponse{Experiment: exp, Config: protoutils.ToStruct(conf)}, nil
 }
 
+func (a *apiServer) DeleteExperiment(
+	_ context.Context, req *apiv1.DeleteExperimentRequest,
+) (*apiv1.DeleteExperimentResponse, error) {
+	expID := int(req.ExperimentId)
+	exp, err := a.m.db.ExperimentByID(expID)
+	switch {
+	case errors.Cause(err) == db.ErrNotFound:
+		return nil, status.Errorf(codes.NotFound, "experiment %d not found", expID)
+	case err != nil:
+		return nil, fmt.Errorf("failed to retrieve experiment: %w", err)
+	}
+
+	if !model.TerminalStates[exp.State] {
+		return nil, fmt.Errorf("cannot delete experiment in %s state", exp.State)
+	}
+
+	switch exists, eErr := a.m.db.ExperimentHasCheckpointsInRegistry(expID); {
+	case eErr != nil:
+		return nil, fmt.Errorf("failed to check model registry for references to %d", expID)
+	case exists:
+		return nil, status.Errorf(codes.InvalidArgument, "checkpoints are registered as model versions")
+	}
+
+	agentUserGroup, err := a.m.db.AgentUserGroup(*exp.OwnerID)
+	switch {
+	case err != nil:
+		return nil, errors.Errorf("cannot find user and group for experiment %v", expID)
+	case agentUserGroup == nil:
+		agentUserGroup = &a.m.config.Security.DefaultTask
+	}
+
+	exp.Config.CheckpointStorage.SaveExperimentBest = 0
+	exp.Config.CheckpointStorage.SaveTrialBest = 0
+	exp.Config.CheckpointStorage.SaveTrialLatest = 0
+	if sErr := a.m.db.SaveExperimentConfig(exp); sErr != nil {
+		return nil, errors.Wrapf(sErr, "failed to patch experiment checkpoint storage %d", exp.ID)
+	}
+	addr := actor.Addr(fmt.Sprintf("delete-checkpoint-gc-%s", uuid.New().String()))
+	if gcErr := a.m.system.MustActorOf(addr, &checkpointGCTask{
+		agentUserGroup: agentUserGroup,
+		taskSpec:       a.m.taskSpec,
+		rm:             a.m.rm,
+		db:             a.m.db,
+		experiment:     exp,
+	}).AwaitTermination(); gcErr != nil {
+		return nil, errors.Wrapf(gcErr, "failed to gc checkpoints for experiment: %d", exp.ID)
```

Comment:
Eh, not sure the errors need to have the experiment ID embedded, since this is in response to a request for a specific experiment anyway.